lifter/OperandUtils.cpp

```diff
@@ -33,7 +121,42 @@ using namespace llvm::PatternMatch;
 
 Value* doPatternMatching(Instruction::BinaryOps I, Value* op0, Value* op1) {
   Value* X = nullptr;
+  Value* Y = nullptr;
+  Value* Z = nullptr;
+
   switch (I) {
+  case Instruction::Add:
+  case Instruction::Or: {
+    Value *Z = nullptr, *A = nullptr, *B = nullptr, *C = nullptr;
+
+    // Match (~A & B) | (A & C)
+    if ((match(op0, m_And(m_Not(m_Value(X)), m_Value(B))) &&
+         match(op1, m_And(m_Value(X), m_Value(C)))) ||
+        (match(op1, m_And(m_Not(m_Value(X)), m_Value(B))) &&
+         match(op0, m_And(m_Value(X), m_Value(C))))) {
+      // This matches (~A & B) | (A & C)
+      // Simplify to A ? C : B
+
+      // X is ( max(v) - v)
+      printvalue(X);
+      printvalue(B);
+      printvalue(C);
+      if (auto X_inst = dyn_cast<Instruction>(X)) {
+        auto pv = GEPStoreTracker::computePossibleValues(X_inst);
+        if (pv.size() == 2) {
+          // check if pv is 0, -1
+
+          IRBuilder<> builder(cast<Instruction>(op0));
+          return createSelectFolder(builder, X, C, B, "selectEZ");
+        }
+      }
+
+      // printvalue2(tryCompute.getMinValue()); if 0
+      // printvalue2(tryCompute.getMaxValue()); if -1
+      // do the simplfiication
+    }
+    break;
+  }
   case Instruction::And: {
     // X & ~X
     // how the hell we remove this zext and truncs it looks horrible
```

Comment:
we have m_TruncOrSelf and m_ZextOrSelf, create m_TruncOrZextOrSelf and rewrite

Reply:
no, we cant match m_TruncOrSelf, because we want to match when its both trunc, so the bits are same.